### PR TITLE
Implement bindings for the Zeebe workflow engine

### LINKED ISSUE #805: Implement bindings for the Zeebe workflow engine
[Zeebe](https://camunda.com/products/cloud/workflow-engine/) is a workflow engine for microservice orchestration. With the bindings, a Dapr user can implement job workers based on the Dapr input binding and it's also possible to interact with the workflow engine based on multiple output bindings.

These are all commands which are currently implemented by Zeebe and which can be triggered by a output binding:

- get topology information
- deploy a workflow
- create an instance of a workflow
- cancel a workflow instance
- set variables for a workflow instance
- resolves an incident
- publish a message
- activate a job
- complete a job
- fail a job
- update the job retries
- throw an error

### PR DESCRIPTION
**Pull request by akkie**
# Description

Implement bindings for the [Zeebe](https://camunda.com/products/cloud/workflow-engine/) workflow engine. This PR implements a job worker input binding that can be used to create workers which will be triggered by Zeebe. The PR implements also an output binding for all the Zeebe commands:

- get topology information
- deploy a workflow
- create an instance of a workflow
- cancel a workflow instance
- set variables for a workflow instance
- resolves an incident
- publish a message
- activate a job
- complete a job
- fail a job
- update the job retries
- throw an error

## Issue reference

Fixes #805 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ https://github.com/dapr/docs/pull/1441

### CODE REVIEW

File: bindings/zeebe/command/activate_jobs.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	"fmt"
+	"time"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/components-contrib/metadata"
+)
+
+const (
+	// errors
+	missingJobTypeErrorMsg           = "jobType is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/command/activate_jobs.go:21]: It's better to have these as error variables. Example:

```golang
var (
    ErrMissingJobType           = errors.New("jobType is a required attribute")
    ErrMissingMaxJobsToActivate = errors.New("maxJobsToActivate is a required attribute")
)
```

File: bindings/zeebe/client.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/dapr/pkg/logger"
+	"github.com/zeebe-io/zeebe/clients/go/pkg/zbc"
+)
+
+const (
+	// errors
+	missingGatewayAddrErrorMsg = "gatewayAddr is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/client.go:20]: Please use error variables with a comment if exposed by the package.

```golang
var (
	ErrMissingGatewayAddr = errors.New("gatewayAddr is a required attribute")
)
```

File: bindings/zeebe/command/activate_jobs.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	if payload.FetchVariables != nil {
+		cmd = cmd.FetchVariables(payload.FetchVariables...)
+	}
+
+	ctx := context.Background()
+	response, err := cmd.Send(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot activate jobs for type %s: %s", payload.JobType, err)
```
Comment (pkedy) [inline bindings/zeebe/command/activate_jobs.go:67]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/activate_jobs.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	jsonResponse, err := json.Marshal(response)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal response to json: %s", err)
```
Comment (pkedy) [inline bindings/zeebe/command/activate_jobs.go:72]: `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/cancel_instance.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/dapr/components-contrib/bindings"
+)
+
+const (
+	// errors
+	missingWorkflowInstanceKeyErrorMsg = "workflowInstanceKey is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/command/cancel_instance.go:19]: Follow same pattern of using error variables above

File: bindings/zeebe/command/complete_job.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	if payload.Variables != nil {
+		cmd3, err = cmd2.VariablesFromObject(payload.Variables)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	_, err = cmd3.Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot complete job for key %d: %s", payload.JobKey, err)
```
Comment (pkedy) [inline bindings/zeebe/command/complete_job.go:46]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/create_instance.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/zeebe-io/zeebe/clients/go/pkg/commands"
+)
+
+const (
+	// errors
+	ambiguousCreationVarsErrorMsg = "either 'bpmnProcessId' or 'workflowKey' must be passed, not both at the same time"
```
Comment (pkedy) [inline bindings/zeebe/command/create_instance.go:20]: Use error variable

File: bindings/zeebe/command/create_instance.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	if payload.Variables != nil {
+		cmd3, err = cmd3.VariablesFromObject(payload.Variables)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	response, err := cmd3.Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot create instane for %s: %s", errorDetail, err)
```
Comment (pkedy) [inline bindings/zeebe/command/create_instance.go:67]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/create_instance.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	jsonResponse, err := json.Marshal(response)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal response to json: %s", err)
```
Comment (pkedy) [inline bindings/zeebe/command/create_instance.go:72]: `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/deploy_workflow.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/zeebe-io/zeebe/clients/go/pkg/pb"
+)
+
+const (
+	// metadata
+	fileName = "fileName"
+	fileType = "fileType"
+
+	// errors
+	missingFileNameErrorMsg = "fileName is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/command/deploy_workflow.go:25]: Use error variables

File: bindings/zeebe/command/deploy_workflow.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		resourceType = pb.WorkflowRequestObject_FILE
+	} else {
+		return nil, errors.New(invalidFileTypeErrorMsg)
+	}
+
+	response, err := z.client.NewDeployWorkflowCommand().
+		AddResource(req.Data, deployFileName, resourceType).
+		Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot deploy workflow with fileName %s: %s", deployFileName, err)
```
Comment (pkedy) [inline bindings/zeebe/command/deploy_workflow.go:68]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/fail_job.go [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+		JobKey(*payload.JobKey).
+		Retries(*payload.Retries)
+
+	if payload.ErrorMessage != "" {
+		cmd = cmd.ErrorMessage(payload.ErrorMessage)
+	}
+
+	_, err = cmd.Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot fail job for key %d: %s", payload.JobKey, err)
```
Comment (pkedy) [inline bindings/zeebe/command/fail_job.go:53]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/publish_message.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	"fmt"
+	"time"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/components-contrib/metadata"
+)
+
+const (
+	// errors
+	missingMessageNameErrorMsg = "messageName is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/command/publish_message.go:21]: Use error variable

File: bindings/zeebe/command/publish_message.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+		cmd, err = cmd.VariablesFromObject(payload.Variables)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx := context.Background()
+	response, err := cmd.Send(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot publish message with name %s: %s", payload.MessageName, err)
```
Comment (pkedy) [inline bindings/zeebe/command/publish_message.go:65]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/publish_message.go [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	jsonResponse, err := json.Marshal(response)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal response to json: %s", err)
```
Comment (pkedy) [inline bindings/zeebe/command/publish_message.go:70]: `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/resolve_incident.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/dapr/components-contrib/bindings"
+)
+
+const (
+	// errors
+	missingIncidentKeyErrorMsg = "incidentKey is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/command/resolve_incident.go:19]: Use error variable

File: bindings/zeebe/command/resolve_incident.go [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+	if payload.IncidentKey == nil {
+		return nil, errors.New(missingIncidentKeyErrorMsg)
+	}
+
+	_, err = z.client.NewResolveIncidentCommand().
+		IncidentKey(*payload.IncidentKey).
+		Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot resolve incident for key %d: %s", payload.IncidentKey, err)
```
Comment (pkedy) [inline bindings/zeebe/command/resolve_incident.go:41]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/set_variables.go [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/dapr/components-contrib/bindings"
+)
+
+const (
+	// errors
+	missingElementInstanceKeyErrorMsg = "elementInstanceKey is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/command/set_variables.go:19]: Use error variables

File: bindings/zeebe/command/set_variables.go [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+	cmd, err := z.client.NewSetVariablesCommand().
+		ElementInstanceKey(*payload.ElementInstanceKey).
+		VariablesFromObject(payload.Variables)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := cmd.Local(payload.Local).Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot set variables for element instance key %d: %s", payload.ElementInstanceKey, err)
```
Comment (pkedy) [inline bindings/zeebe/command/set_variables.go:53]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/throw_error.go [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+		JobKey(*payload.JobKey).
+		ErrorCode(payload.ErrorCode)
+
+	if payload.ErrorMessage != "" {
+		cmd = cmd.ErrorMessage(payload.ErrorMessage)
+	}
+
+	_, err = cmd.Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot throw error for job key %d: %s", payload.JobKey, err)
```
Comment (pkedy) [inline bindings/zeebe/command/throw_error.go:53]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/throw_error.go [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/dapr/components-contrib/bindings"
+)
+
+const (
+	// errors
+	missingErrorCodeErrorMsg = "errorCode is a required attribute"
```
Comment (pkedy) [inline bindings/zeebe/command/throw_error.go:19]: Use error variable

File: bindings/zeebe/command/topology.go [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/dapr/components-contrib/bindings"
+)
+
+func (z *ZeebeCommand) topology() (*bindings.InvokeResponse, error) {
+	response, err := z.client.NewTopologyCommand().Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot get zeebe toplogy: %s", err)
```
Comment (pkedy) [inline bindings/zeebe/command/topology.go:19]: `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/topology.go [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	jsonResponse, err := json.Marshal(response)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal response to json: %s", err)
```
Comment (pkedy) [inline bindings/zeebe/command/topology.go:24]: `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/command/update_job_retries.go [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+
+	cmd1 := z.client.NewUpdateJobRetriesCommand().JobKey(*payload.JobKey)
+	var cmd2 commands.DispatchUpdateJobRetriesCommand = cmd1
+	if payload.Retries != nil {
+		cmd2 = cmd1.Retries(*payload.Retries)
+	}
+
+	_, err = cmd2.Send(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("cannot uodate job retries for key %d: %s", payload.JobKey, err)
```
Comment (pkedy) [inline bindings/zeebe/command/update_job_retries.go:42]: Last `%s` should be `%w` per Go 1.13 error wrapping

File: bindings/zeebe/jobworker/jobworker.go [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+	if err != nil {
+		h.failJob(client, job, err.Error())
+
+		return
+	}
+
+	variablesMap := make(map[string]interface{})
+	err = json.Unmarshal(resultVariables, &variablesMap)
+	if err != nil {
+		h.failJob(client, job, fmt.Sprintf("Cannot parse variables from binding result %s; got error %s", string(resultVariables), err.Error()))
```
Comment (pkedy) [inline bindings/zeebe/jobworker/jobworker.go:149]: Last `%s` should be `%w` per Go 1.13 error wrapping

**Review by pkedy — COMMENTED** [review]
Just a first pass before the weekend :)

Noticed some tweaks to be made w/r/t errors

- Use error variables
- Instead if using `%s` for an error in `fmt.Errorf`, use `%v` or `%w` if you want to wrap the error. https://blog.golang.org/go1.13-errors

Also please consider using interfaces to Mock instead of `github.com/stretchr/testify/mock`.

https://github.com/dapr/components-contrib/blob/master/bindings/twilio/sms/sms_test.go#L21-L38

File: bindings/zeebe/command/activate_jobs.go (hunk 1 above)
Comment (akkie) [reply bindings/zeebe/command/activate_jobs.go]: Fixed

**Comment by akkie** [comment]
@pkedy Thanks for your suggestions. I've resolved all your suggestions regarding the error constants and also regarding the error wrapping.

Regarding your suggestion with the mocks. This is my first contribution in go. Do you have a tutorial for me which describes this way of mocking?

**Comment by pkedy** [comment]
Hi @akkie. Here is a good example here.

https://github.com/dapr/dapr/blob/master/pkg/actors/actors_test.go#L39-L57

Basic structure is the mock struct embeds the interface is it mocking. This enables you to only have to mock the methods that are required by the test yet the struct still implements the interface (because it is embedded). Of course if the test ends up calling an unimplemented method, it panics, but then you know to implement it. In this example, a channel which would be more challenging when using a mocking library.

**Comment by akkie** [comment]
@pkedy I have refactored the test for the cancel_instance command. Could you please have a look if this is ok before I start to refactor the other tests.

File: bindings/zeebe/command/cancel_instance_test.go [hunk 26]
```diff
@@ -59,20 +69,12 @@ func TestCancelInstance(t *testing.T) {
...
 		req := &bindings.InvokeRequest{Data: data, Operation: cancelInstanceOperation}
 
-		mc := new(MockClient)
-		cmd1 := new(MockCancelInstanceCommandStep1)
-		cmd2 := new(MockCancelInstanceCommandStep2)
-
-		mc.On("NewCancelInstanceCommand").Return(cmd1)
-		cmd1.On("WorkflowInstanceKey", *payload.WorkflowInstanceKey).Return(cmd2)
-		cmd2.On("Send", mock.AnythingOfType("*context.emptyCtx")).Return(new(pb.CancelWorkflowInstanceResponse), nil)
+		mc := new(mockClient)
```
Comment (pkedy) [inline bindings/zeebe/command/cancel_instance_test.go:72]: nit: could just be `var mc mockClient`.

**Comment by pkedy** [comment]
@akkie Yes! I put 1 small nit there but this is exactly the pattern for mocking via interfaces.

File: bindings/zeebe/command/cancel_instance_test.go (hunk 26 above)
Comment (akkie) [reply bindings/zeebe/command/cancel_instance_test.go:72]: If I do that, then I get the error message:

```
cannot use mc (type mockClient) as type zbc.Client in field value:
	mockClient does not implement zbc.Client (NewCancelInstanceCommand method has pointer receiver)
```

Comment (pkedy) [reply bindings/zeebe/command/cancel_instance_test.go]: Then you "share" the pointer to `mc` using `&`.

**Comment by pkedy** [comment]
@akkie please ping me when you want me to take another look at this. Thanks!

**Comment by akkie** [comment]
@pkedy From my perspective the PR is finished. I've also finished the documentation.

I'm not sure why the build failed. Seems like an issue with the build agent. Could you please trigger the pipelines again?

**Comment by pkedy** [comment]
@akkie We very recently moved the logger package from `dapr/dapr` to `dapr/kit` so I fixed that for you and made a few miscellaneous tweaks.

**Review by pkedy — APPROVED** [review]
LGTM but I did not test with the Zeebe product. I'd recommend a follow up PR that add integration tests using your [docker image](https://camunda.com/download/). We have some examples of that in our conformance tests in dapr/components-contrib.

**Comment by akkie** [comment]
@pkedy I will create a follow up PR for integration testing. If you like to play with Zeebe, I've setup a small example application and a description how Zeebe can be setup locally.

https://github.com/PlanBGmbH/zeebe-dapr-example

**Review by artursouza — APPROVED** [review]

**Pull request merged**